feat: string and password import with settings

vfoucault opened the pull request:
This PR add the ability to import `random_string` and `random_password` with their settings. 
Settings might be specified in the resource address while importing and alter a set of default settings.

This is to avoid the re-creation of random string/password after import, settings being `null`.

Despite [this workaround](https://github.com/hashicorp/terraform-provider-random/issues/106#issuecomment-692568119), I believe that being able to specify  resource settings is a better approach.

This PR resolves #106 #119 

A resource `address` could be either a single string, then default settings are applied or a string with a set of comma separated key=value

eg:

```
bladibla length=8,special=false
```

will import the resource with `result` = `bladibla` and `length` of `8` and `special` set to `false`

Full example:

```hcl
resource "random_password" "test" {
  keepers = {
    bla: "dibla"
    key: "value"
  }
  length           = 6
  special          = false
  upper            = true
  lower            = true
  number           = true
  min_numeric      = 0
  min_upper        = 0
  min_lower        = 0
  min_special      = 0
  override_special = ""
}
```

```
~# terraform import random_password.test "234567 length=6,special=false,keepers={\"bla\":\"dibla\",\"key\":\"value\"}"                                                                                         
random_password.test: Importing from ID "234567 length=6,special=false,keepers={\"bla\":\"dibla\",\"key\":\"value\"}"...
random_password.test: Import prepared!
  Prepared random_password for import
random_password.test: Refreshing state... [id=none]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

```

Then with `terraform plan`

```
~# terraform plan                                                                                                                                                                                      
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

random_password.test: Refreshing state... [id=none]

------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.

```

vfoucault commented:
Tests are randomly failing for no obvious reason. Even though, they are all passing

vfoucault commented:
Whether you want to give a try to this PR, this provider has been published on the registry.

```hcl
terraform {
  required_providers {
    random = {
      source = "ContentSquare/random"
      version = "3.1.0"
    }
  }
}
```

vfoucault commented:
@appilon @kmoe did you had a chance to take a look a this by chance ? kr.

r0bnet commented:
Any update on this? Would make it much easier for us to switch from `random_string` to `random_password` in some of our code.
@vfoucault could you be so kind to resolve the conflicts?

vfoucault commented:
> Any update on this? Would make it much easier for us to switch from `random_string` to `random_password` in some of our code.
> @vfoucault could you be so kind to resolve the conflicts?

Done. I hope that I've done the doc thingy the right way.

tun0 commented on examples/resources/random_string/import.sh:
```diff
@@ -1,2 +1,13 @@
+# Strings can be imported by just specifying the value of the string alongside the resource attributes.
+terraform import random_string.test "test length=4"
+
+# Random Password can be imported by specifying the value of the string alongside the resource attributes.
+terraform import random_string.test "test keepers=nil,length=4,special=true,upper=true,lower=true,number=true,min_numeric=0,min_upper=0,min_lower=0,min_special=0,override_special=_%@"
```
Is this about random strings or passwords :thinking:

vfoucault replied:
Actually both, since the change has been made for string and passwords.

tun0 replied:
It seems the context selection didn't go well. I meant to highlight the difference between the comment and the actual command.

vfoucault replied:
Copy paste issue, i'll fix this

vfoucault commented:
@appilon @kmoe did you had some time to review this pr  by any chance ? thanks in advance.

PVandenheede commented:
Any chance for a review of this PR? Would be nice to have this.

jspiro commented:
Need this!

burton1982 reviewed (approved):
Nice!

nasonfish commented on internal/provider/string.go:
```diff
@@ -219,7 +225,88 @@ func importStringFunc(sensitive bool) schema.StateContextFunc {
+			// Keep default behavior but add default values
+			d.Set("result", val)
+			log.Println(fmt.Sprintf("[DEBUG] using default settings: %v", settings))
+			for key, value := range settings {
```
It looks like this loop includes `length` from the map above, though that might not always match the length of `val`. Just curious, would it be reasonable to calculate the length of `val` and use that as the default here, rather than using 16 if a string is imported with no length specified?

vfoucault replied:
Hi there, I actually thought about it, but this causes issues with passwords that do not match the length specified in the resource definition.

eg. when you specify a length = 8 and importing 32 chars password, then the state of the resource will be 32 and not 8, causing a resource re recreation. 

Length in the import statement must match the length in the resource definition, but password length could be different.

ppihus commented:
Is there something still blocking this PR?
We have a ongoing refactoring with hundreds of imports and this is really blocking us atm.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-random?pullRequest=135)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-random?pullRequest=135)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-random?pullRequest=135).</sub>

wszychta commented:
This PR is really important! It should be merged long time ago.

tolik-u commented:
![image](https://user-images.githubusercontent.com/16921786/162938645-3fe1dfb7-247b-4d67-a542-00b6ab78f35f.png)
I can't even

vfoucault commented:
I've rebased onto main because of the merge conflcts, updated accordingly the tests and updated according to new standards.

```
 make testacc
TF_ACC=1 go test -v -cover -timeout 120m ./...
?       github.com/terraform-providers/terraform-provider-random        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccResourceID
--- PASS: TestAccResourceID (1.05s)
=== RUN   TestAccResourceID_importWithPrefix
--- PASS: TestAccResourceID_importWithPrefix (0.85s)
=== RUN   TestAccResourceIntegerBasic
=== PAUSE TestAccResourceIntegerBasic
=== RUN   TestAccResourceIntegerUpdate
=== PAUSE TestAccResourceIntegerUpdate
=== RUN   TestAccResourceIntegerSeedless_to_seeded
=== PAUSE TestAccResourceIntegerSeedless_to_seeded
=== RUN   TestAccResourceIntegerSeeded_to_seedless
=== PAUSE TestAccResourceIntegerSeeded_to_seedless
=== RUN   TestAccResourceIntegerBig
=== PAUSE TestAccResourceIntegerBig
=== RUN   TestAccResourcePasswordBasic
--- PASS: TestAccResourcePasswordBasic (0.89s)
=== RUN   TestAccResourcePasswordOverride
--- PASS: TestAccResourcePasswordOverride (0.70s)
=== RUN   TestAccResourcePasswordMin
--- PASS: TestAccResourcePasswordMin (0.75s)
=== RUN   TestAccResourcePet_basic
--- PASS: TestAccResourcePet_basic (0.70s)
=== RUN   TestAccResourcePet_length
--- PASS: TestAccResourcePet_length (0.69s)
=== RUN   TestAccResourcePet_prefix
--- PASS: TestAccResourcePet_prefix (0.69s)
=== RUN   TestAccResourcePet_separator
--- PASS: TestAccResourcePet_separator (0.70s)
=== RUN   TestAccResourceShuffleDefault
--- PASS: TestAccResourceShuffleDefault (0.66s)
=== RUN   TestAccResourceShuffleShorter
--- PASS: TestAccResourceShuffleShorter (0.71s)
=== RUN   TestAccResourceShuffleLonger
--- PASS: TestAccResourceShuffleLonger (0.68s)
=== RUN   TestAccResourceShuffleEmpty
--- PASS: TestAccResourceShuffleEmpty (0.68s)
=== RUN   TestAccResourceShuffleOne
--- PASS: TestAccResourceShuffleOne (0.67s)
=== RUN   TestResourceStringMigrateState
--- PASS: TestResourceStringMigrateState (0.00s)
=== RUN   TestAccResourceString
--- PASS: TestAccResourceString (0.88s)
=== RUN   TestAccResourceStringOverride
--- PASS: TestAccResourceStringOverride (0.68s)
=== RUN   TestAccResourceStringMin
--- PASS: TestAccResourceStringMin (0.69s)
=== RUN   TestAccResourceStringErrors
--- PASS: TestAccResourceStringErrors (0.43s)
=== RUN   TestAccResourcePassword_import_simple
=== PAUSE TestAccResourcePassword_import_simple
=== RUN   TestAccResourcePassword_import_stronger
=== PAUSE TestAccResourcePassword_import_stronger
=== RUN   TestAccResourceUUID
--- PASS: TestAccResourceUUID (0.85s)
=== CONT  TestAccResourceIntegerBasic
=== CONT  TestAccResourceIntegerBig
=== CONT  TestAccResourceIntegerSeeded_to_seedless
=== CONT  TestAccResourcePassword_import_stronger
=== CONT  TestAccResourcePassword_import_simple
=== CONT  TestAccResourceIntegerUpdate
=== CONT  TestAccResourceIntegerSeedless_to_seeded
--- PASS: TestAccResourcePassword_import_stronger (1.70s)
--- PASS: TestAccResourcePassword_import_simple (1.70s)
--- PASS: TestAccResourceIntegerBig (1.82s)
--- PASS: TestAccResourceIntegerBasic (1.84s)
--- PASS: TestAccResourceIntegerSeeded_to_seedless (2.16s)
--- PASS: TestAccResourceIntegerSeedless_to_seeded (2.24s)
--- PASS: TestAccResourceIntegerUpdate (2.24s)
PASS
coverage: 85.6% of statements
ok      github.com/terraform-providers/terraform-provider-random/internal/provider      17.225s coverage: 85.6% of statements
```

bendbennett commented:
Hi @vfoucault, thank you for submitting the PR and apologies in the delay in us getting to it.

Your proposal is an interesting one, as you're essentially overloading the import `result` with additional attributes/fields for use in populating the resource during import. There are a number of [use cases described for this scenario](https://github.com/hashicorp/terraform-providers-devex/issues/2).

At this time, we are intending to implement [Import via Configuration](https://github.com/hashicorp/terraform-providers-devex/issues/2) for providers developed with the [Framework](https://github.com/hashicorp/terraform-plugin-framework) to deliver the ability to import multiple attributes/fields. We would therefore like to avoid "overloading" the import id (or result in this case) for this purpose.

MarkRKar commented:
> This is to avoid the re-creation of random string/password after import, settings being null.

I think this fixes it https://github.com/hashicorp/terraform-provider-random/pull/256

simonweil commented:
I'm still having this issue with 3.6
@bendbennett Why not merge this as a workaround until the framework supports the configuration import?
